### ISSUE TITLE
Macro calls in enum

### ISSUE THREAD

**Issue by kirbyfan64** [crystal-lang/crystal#3268]
Related to #424.

To my dismay, I realized I couldn't do this:

```crystal
macro f(id, value)
  {{id}} = {{value}}
end

enum X
  f A, 1
end
```

Is there a reason stuff like this isn't supported?

**Comment by asterite** [comment]
Normal expressions like calls, assigns, etc., are not being parsed right now inside enums. Maybe they should, and this will start working.

I think the reason for this is that in the beginning one could define enums just by defining the constants, nothing else was allowed. Then we decided that methods could be defined. And then class vars. So in the end almost all of the grammar should be allowed.

**Comment by kirbyfan64** [comment]
Huh...how hard do you think this would be? I just edited the parser and got the example to compile, but I feel like it was too easy...

**Comment by asterite** [comment]
I'm not sure, what happens if you define a constant `A = "hello"` like that, using a macro call?

**Comment by kirbyfan64** [comment]
@asterite It just defines it as a normal constant. Is that supposed to define it like an enum value, as if there were no macro?

**Comment by asterite** [comment]
Yes, constant definitions inside enums are treated in a different way, and they are always checked to be integers. It shouldn't be hard to implement this, I'll try to do it before 0.19.1.

**Comment by kirbyfan64** [comment]
Just ran into this again. :(

**Comment by bew** [comment]
Just stepped into this but for a different reason:

I want to add a method to some specific enums, as we currently can't include a module in an enum, I tried with a macro and it didn't worked either:
```cr
macro define_validator
  def valid?
    value <= {{ @type.constants.last }}.value
  end
end

enum Foo
  A
  B
  
  define_validator  # unexpected token: define_validator
end

enum Bar
  C
  D
  
  define_validator  # unexpected token: define_validator
end
```
https://carc.in/#/r/3mjj

I could add this method to `Enum` directly, but it may not work with all enums..

**Comment by willhbr** [comment]
I also just ran into this same limitation trying to include a module in an enum, it'd be great if enums matched what you can do with classes and structs (as much as is possible)!
